centaurCwlRunner/src/main/scala/centaur/cwl/Outputs.scala

```diff
@@ -0,0 +1,106 @@
+package centaur.cwl
+
+import cats.effect.IO
+import centaur.api.CentaurCromwellClient
+import cromwell.api.model.SubmittedWorkflow
+import cwl.{CwlDecoder, CwlType, MyriadOutputType, File => CwlFile}
+import io.circe.Json
+import io.circe.generic.auto._
+import io.circe.literal._
+import io.circe.refined._
+import io.circe.shapes._
+import io.circe.syntax._
+import shapeless.{Inl, Poly1}
+import spray.json.{JsNumber, JsObject, JsString, JsValue}
+
+import scalaz.syntax.std.map._
+
+object Outputs {
+
+  //When the string returned is not valid JSON, it is effectively an exception as CWL runner expects JSON to be returned
+  def handleOutput(submittedWorkflow: SubmittedWorkflow): String = {
+    val metadata: Map[String, JsValue] = CentaurCromwellClient.metadata(submittedWorkflow).get.value
+
+    //Sorry for all the nesting, but spray json JsValue doesn't have optional access methods like Argonaut/circe,
+    //thus no for comprehensions for us :(
+    metadata.get("submittedFiles.workflow") match {
+      case Some(JsString(some)) =>
+
+        val cwl = CwlDecoder.decodeTopLevelCwl(some)
+
+        cwl.value.attempt.unsafeRunSync() match {
+          case Right(Right(cwl)) =>
+
+            CentaurCromwellClient.outputs(submittedWorkflow).get.outputs match {
+              case JsObject(map) =>
+                val typeMap: Map[String, MyriadOutputType] = cwl.fold(CwlOutputsFold)
+                val mungeTypeMap = typeMap.mapKeys(stripTypeMapKey)
+
+                val mungeOutputMap = map.mapKeys(stripOutputKey)
+
+                mungeOutputMap.
+                  //This lets us operate on the values of the output values and types for a particular output key
+                  intersectWith(mungeTypeMap)(resolveOutput).
+                  //converting the whole response to Json using Circe's auto-encoder derivation
+                  asJson.
+                  //drop null values so that we don't print when Option == None
+                  pretty(io.circe.Printer.spaces2.copy(dropNullValues = true))
+              case other => s"it seems cromwell is not returning outputs as a Jsobject but is instead a $other"
+            }
+          case Right(Left(error)) => s"couldn't parse $some: $error"
+          case Left(error) => s"couldn't parse $some: $error"
```

Comment:
What's the difference between a `Right(Left(error))` and a `Left(error)` ? Do they really both mean parsing failed or do they have a different cause ?

Reply:
the outer is a result of running `IO.attempt` which catches any exceptions.

The inner is a result of a parsing error, i.e. non-exception. 

I see I left some copypasta here and should fix the error message.